[AOMP][ASan] Major Refactoring of AOMP scripts for ASan support.

ampandey-AMD opened the pull request:
There is a requirement from customers to ship ASan instrumented version of OpenMP libraries i.e OpenMP{Release(llvm/lib),Debug(llvm/lib-debug),Perf(llvm/lib-perf)}. The current patch achieves the following:-

1. Build & Install support OpenMP{Release-ASan(llvm/lib/asan)}.
2. Build & Install support OpenMP{Debug-ASan(llvm/lib-debug/asan)}.
3. Build & Install support OpenMP{Perf-ASan(llvm/lib-perf/asan)}.
4. AOMP_BUILD_SANITIZER=1 (Default for AOMP Standalone builds).

ampandey-AMD committed:
[AOMP][ASan] Major Refactoring of AOMP scripts for ASan support.

There is a requirement from customers to ship ASan instrumented version of OpenMP libraries i.e
OpenMP{Release(llvm/lib),Debug(llvm/lib-debug),Perf(llvm/lib-perf)}.
The current patch achieves the following:-

1. Build & Install support OpenMP{Release-ASan(llvm/lib/asan)}.
2. Build & Install support OpenMP{Debug-ASan(llvm/lib-debug/asan)}.
3. Build & Install support OpenMP{Perf-ASan(llvm/lib-perf/asan)}.
4. AOMP_BUILD_SANITIZER=1 (Default for AOMP Standalone builds).

ampandey-AMD committed:
[AOMP][ASan] Major Refactoring of AOMP scripts for ASan support.

There is a requirement from customers to ship ASan instrumented version of OpenMP libraries i.e
OpenMP{Release(llvm/lib),Debug(llvm/lib-debug),Perf(llvm/lib-perf)}.
The current patch achieves the following:-

1. Build & Install support OpenMP{Release-ASan(llvm/lib/asan)}.
2. Build & Install support OpenMP{Debug-ASan(llvm/lib-debug/asan)}.
3. Build & Install support OpenMP{Perf-ASan(llvm/lib-perf/asan)}.
4. AOMP_BUILD_SANITIZER=1 (Default for AOMP Standalone builds).

ronlieb reviewed (changes requested):
i applied the patch, and also enabled lib-perf build by default.  
i see the directories for lib/asan and lib-debug/asan   
but i dont see the directory lib-perf/asan

ampandey-AMD commented:
> i applied the patch, and also enabled lib-perf build by default. i see the directories for lib/asan and lib-debug/asan but i dont see the directory lib-perf/asan

Yes there is missing part of code in llvm-project/openmp/CMakelists.txt.

We have to modify below this part of code 
` if(SANITIZER_AMDGPU)
  if(CMAKE_BUILD_TYPE MATCHES Debug)
    set(OPENMP_LIBDIR_SUFFIX "-debug/asan")
    set(OPENMP_INSTALL_LIBDIR "lib${OPENMP_LIBDIR_SUFFIX}")
  else()
    set(OPENMP_LIBDIR_SUFFIX "/asan")
    set(OPENMP_INSTALL_LIBDIR "lib${OPENMP_LIBDIR_SUFFIX}")
  endif()
endif()
`

with this code 

`if(SANITIZER_AMDGPU)
  if(CMAKE_BUILD_TYPE MATCHES Debug)
    set(OPENMP_LIBDIR_SUFFIX "-debug/asan")
    set(OPENMP_INSTALL_LIBDIR "lib${OPENMP_LIBDIR_SUFFIX}")
  else()
    if(${LIBOMPTARGET_PERF})
       set(OPENMP_LIBDIR_SUFFIX "-perf/asan")
       set(OPENMP_INSTALL_LIBDIR "lib${OPENMP_LIBDIR_SUFFIX}")
    else()
       set(OPENMP_LIBDIR_SUFFIX "/asan")
       set(OPENMP_INSTALL_LIBDIR "lib${OPENMP_LIBDIR_SUFFIX}")
  endif()
endif()
`

But gerrit is not working for me If it is working for you can you please make that corresponding change in the CMakeLists.txt?

Thanks 
Amit

ronlieb commented:
seeing a lot of test failures with a build using this patch
=== omp_hip
-- openmpi found at /opt/openmpi-4.1.5
-- openmpi found at /opt/openmpi-4.1.5
/home/rlieberm/rocm/aomp_17.0-5/bin/clang++ -std=c++11 -O3  -fopenmp -x hip -L /home/rlieberm/rocm/aomp_17.0-5/hip -L /home/rlieberm/rocm/aomp_17.0-5/lib -lamdhip64 -Wl,-rpath,/home/rlieberm/rocm/aomp_17.0-5/lib --offload-arch=gfx90a  matrixmul.cpp -o matrixmul
-- openmpi found at /opt/openmpi-4.1.5
./matrixmul
==2617513==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
make: *** [Makefile:65: run] Error 1
-- openmpi found at /opt/openmpi-4.1.5

Known Fails: aomp_hip_launch_test matmul_hip_omp_printf_fails matrixmul_omp_target

b-sumner commented:
That failure indicates an instrumented library was linked in but the main program wasn't instrumented.  I don't see -fsanitize=address on the clang++ command line, so I'd guess the wrong runtime was linked.

ronlieb commented:
we do now seem to be picking up  libclang_rt.asan...
so thats a bug to dig into

libclang_rt.asan-x86_64.so => /work/rlieberm/rocm/aomp_17.0-5/lib/clang/17.0.0/lib/linux/libclang_rt.asan-x86_64.so (0x00007fc6aff88000)

ronlieb committed:
Enable openmp library build:   lib-perf. (#551)

The lib-perf build is built with these additional CMAKE settings  
    -DLIBOMPTARGET_ENABLE_DEBUG=OFF 
    -DCMAKE_BUILD_TYPE=Release 
     -DLIBOMPTARGET_PERF=ON      

 

next steps: 
  work on adding driver flags to simplify selection of library to use in RPATH/RUNPATH

 -fopenmp-librtl={debug,devel,perf}
    default is devel  : could change in the future

-fopenmp-librtl=debug
    Library rpath: [$ORIGIN:$ORIGIN/../lib:$ORIGIN/../lib64:/home/rlieberm/rocm/aomp_17.0-5/lib-debug]

-fopenmp-librtl=devel     # Currently the default 
    Library rpath: [$ORIGIN:$ORIGIN/../lib:$ORIGIN/../lib64:/home/rlieberm/rocm/aomp_17.0-5/lib]

-fopenmp-librtl=perf
    Library rpath: [$ORIGIN:$ORIGIN/../lib:$ORIGIN/../lib64:/home/rlieberm/rocm/aomp_17.0-5/lib-perf]

gregrodgers committed:
we need map(from:new_len) to be consistent with c example

gregrodgers committed:
Merge branch 'aomp-dev' of https://github.com/ROCm-Developer-Tools/aomp into aomp-dev

gregrodgers committed:
temporarily turn off CLANG_INCLUDE_TESTS due to upstream problem with unittests

ampandey-AMD committed:
Merge branch 'asan-aomp-build' of github.com:ampandey-AMD/aomp into asan-aomp-build

ampandey-AMD commented:
I messed up the git history as it dosen't apply cleanly. I have created another pull request so closing this in favour of https://github.com/ROCm-Developer-Tools/aomp/pull/552
